tribits/ci_support/CDashAnalyzeReportRandomFailures.py

```diff
@@ -0,0 +1,339 @@
+
+        buildSummaryCacheFile = buildSummaryCacheDir+"/"+buildId
+        buildSummaryQueryUrl = CDQAR.getCDashBuildSummaryQueryUrl(cdashSiteUrl, buildId)
+        buildConfigOutput = self.downloadBuildSummaryOffCDash(
+          buildSummaryQueryUrl, buildSummaryCacheFile, verbose=printUrlMode =='all',
+          alwaysUseCacheFileIfExists=True)['configure']['output']
+
+        passingSha1Pair = \
+          self.versionInfoStrategy.getTargetTopicSha1s(buildConfigOutput)
+
+        if self.versionInfoStrategy.checkTargetTopicRandomFailure(passingSha1Pair, nonpassingSha1Pairs):
+          print("\n  Found passing sha1 pair, " + str(passingSha1Pair)+\
+                " in set of nonpassing sha1 pairs: \n"+str(nonpassingSha1Pairs))
+
+          randomFailureSummaries.append(
+            RandomFailureSummary(test['buildName'], test['testname'],
+              testHistoryBrowserUrl, passingSha1Pair))
+
+
+    print("\n*** CDash random failure analysis for " +\
+      cdashProjectName+" "+groupName+" from " +dateRangeStr)
+
+    print("Total number of initial failing tests: "+str(len(initialNonpassingTestsLOD))+"\n")
+
+    print("Found random failing tests: "+str(len(randomFailureSummaries))+"\n")
+
+    cdashReportData.htmlEmailBodyTop += \
+      "Found random failing tests: "+str(len(randomFailureSummaries))+"<br>\n"
+
+    if len(randomFailureSummaries) > 0:
+      cdashReportData.globalPass = False
+
+    # Add number of random failing tests 'rft' found to summary data list
+    cdashReportData.summaryLineDataNumbersList.append(
+      "rft="+str(len(randomFailureSummaries)))
+
+    # Add number of initial failing tests 'ift' scanned to summary
+    # data list
+    cdashReportData.summaryLineDataNumbersList.append(
+      "ift="+str(len(initialNonpassingTestsLOD)))
+
+    for summary in randomFailureSummaries:
+      print(str(summary))
+      summary.singleSummaryReporter(cdashReportData)
+
+    summaryLine = CDQAR.getOverallCDashReportSummaryLine(
+      cdashReportData, cdashProjectName+" "+groupName, dateRangeStr)
+    print("\n"+summaryLine)
+
+    # Finish HTML body paragraph
+    cdashReportData.htmlEmailBodyTop += "\n</p>"
+
+    defaultPageStyle = CDQAR.getDefaultHtmlPageStyleStr()
+
+    if writeEmailToFile:
+      print("\nWriting HTML to file: "+writeEmailToFile+" ...")
+      htmlStr = CDQAR.getFullCDashHtmlReportPageStr(cdashReportData,
+        pageTitle=summaryLine, pageStyle=defaultPageStyle)
+      # print(htmlStr)
+      with open(writeEmailToFile, 'w') as outFile:
+        outFile.write(htmlStr)
+
+    if sendEmailTo:
+      htmlStr = CDQAR.getFullCDashHtmlReportPageStr(cdashReportData,
+      pageStyle=defaultPageStyle)
+      for emailAddress in sendEmailTo.split(','):
+        emailAddress = emailAddress.strip()
+        print("\nSending email to '"+emailAddress+"' ...")
+        msg=CDQAR.createHtmlMimeEmail(
+          sendEmailFrom, emailAddress, summaryLine, "", htmlStr)
+        CDQAR.sendMineEmail(msg)
+
```

Comment:
For long functions like this, I sometimes provide a comment indented at the same level as the function prototype with:

```
  # end runDriver()
```

so that you know your context.